cdap-master/src/main/java/co/cask/cdap/data/tools/LogUpgrade.java

```diff
@@ -0,0 +1,148 @@
+import co.cask.tephra.DefaultTransactionExecutor;
+import co.cask.tephra.TransactionAware;
+import co.cask.tephra.TransactionExecutor;
+import co.cask.tephra.TransactionExecutorFactory;
+import com.google.common.base.Supplier;
+import com.google.common.base.Throwables;
+import com.google.inject.Injector;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.net.URI;
+import java.util.Arrays;
+import java.util.Map;
+
+/**
+ * Upgrades the Logs
+ */
+public class LogUpgrade extends AbstractUpgrade implements Upgrade {
+
+  private static final Logger LOG = LoggerFactory.getLogger(LogUpgrade.class);
+  private static final String[] OTHERS = {"logs", Constants.SYSTEM_NAMESPACE, Constants.DEFAULT_NAMESPACE};
+
+  private final Transactional<DatasetContext<OrderedTable>, OrderedTable> logMDS;
+
+  public LogUpgrade() {
+    this.logMDS = Transactional.of(
+      new TransactionExecutorFactory() {
+        @Override
+        public TransactionExecutor createExecutor(Iterable<TransactionAware> txAwares) {
+          return new DefaultTransactionExecutor(txClient, txAwares);
+        }
+      },
+      new Supplier<DatasetContext<OrderedTable>>() {
+        @Override
+        public DatasetContext<OrderedTable> get() {
+          try {
+            return DatasetContext.of(new LogSaverTableUtil(nonNamespaedFramework, cConf).getMetaTable());
+          } catch (Exception e) {
+            throw Throwables.propagate(e);
+          }
+        }
+      });
+  }
+
+  @Override
+  public void upgrade(Injector injector) throws Exception {
+    final FileMetaDataManager fileMetaDataManager = new FileMetaDataManager(new LogSaverTableUtil(
+      nonNamespaedFramework, cConf), txClient, locationFactory);
+    logMDS.execute(new TransactionExecutor.Function<DatasetContext<OrderedTable>, Void>() {
+      @Override
+      public Void apply(DatasetContext<OrderedTable> ctx) throws Exception {
+        Scanner rows = ctx.get().scan(null, null);
+        Row row;
+        while ((row = rows.next()) != null) {
+          String key = Bytes.toString(Arrays.copyOfRange(row.getRow(), FileMetaDataManager.ROW_KEY_PREFIX.length,
+                                                         row.getRow().length)).trim();
+          Map<byte[], byte[]> allColumns = row.getColumns();
+          for (byte[] startTime : allColumns.keySet()) {
+            long startTimeMs = Bytes.toLong(startTime);
+            String oldPath = Bytes.toString(allColumns.get(startTime));
+            String newPath;
+            String newKey;
+            if (key.startsWith(Constants.Logging.SYSTEM_NAME) || key.startsWith(DEVELOPER_STRING)) {
+              newPath = upgradePath(key, oldPath);
+              newKey = upgradeKey(key);
+              renameLocation(new URI(oldPath).resolve("."), new URI(newPath).resolve("."));
+              fileMetaDataManager.writeMetaData(newKey, startTimeMs, locationFactory.create(new URI(newPath)));
```

Comment:
this and the previous line are not atomic. We should be sure to handle the case where rename location goes through but this does not. Perhaps check the paths before actually renaming?